### PR TITLE
Optimize Simulation

### PR DESCRIPTION
**Pull request by PeterLombaers**
This pull request tries to make a number of methods that are used in simulations faster. It also changes the way the simulation reviewer works. Now it no longer reads and writes data from/to the state at every iterations, but only at intervals or at the end of the simulation. This makes simulations a lot faster to run.

### CODE REVIEW

**Review by J535D165 — COMMENTED** [review]
It looks like a very nice refactor of the simulate class. Happy to merge. 

We might face some challenges with resolving the merge conflict introduced by the creation of the simulation feature in the front-end. See this blame: https://github.com/asreview/asreview/blame/development-v1/asreview/entry_points/simulate.py. I'm having a look at it.

**Comment by J535D165** [comment]
Made an attempt to resolve the merge conflicts. Hopefully, the tests will pass. The work of @terrymyc that introduced the conflict, might implement an almost identical feature. However, I'm not 100% sure about that. Maybe someone wants to have a look.

**Comment by J535D165** [comment]
Thanks a lot, @PeterLombaers. The PR looks good to me. I haven't tested it extensively but will merge and test it in the upcoming week. I'm working on a ProjectAPI that needs to change some things regarding the state API. This will prevent merge conflicts. Hope you can review that PR :)

File: asreview/state/sql_converter.py [hunk 1]
```diff
@@ -207,6 +207,7 @@ def convert_json_settings_metadata(fp, json_fp):
         data_dict['state_version'] = SQLSTATE_VERSION
         data_dict['software_version'] = json_state._state_dict[
             'software_version']
+        data_dict['model_has_trained'] = True
```
[the remark this answers is not in the document]
Comment (PeterLombaers) [reply asreview/state/sql_converter.py:209]: In the state, to check if a model had finished training (the first model after the priors were added), it went to the last_ranking_table to check if it was still empty. This seemed a bit expensive though, so I changed it into a boolean value 'model_has_trained' in the settings.

**Pull request merged**